pkg/cmd/cli/cmd/startbuild.go

```diff
@@ -488,36 +491,78 @@ func streamPathToBuild(repo git.Repository, in io.Reader, out io.Writer, client
 				glog.V(6).Infof("Unable to read Git info from %q: %v", clean, gitErr)
 			}
 
+			// In case we create a separate directory for transferring a Git repo,
+			// we need to have a channel ot make sure we only remove it when the
+			// transfer is finished
+			transferFinished := make(chan bool)
+
 			if asRepo {
 				fmt.Fprintf(out, "Uploading %q at commit %q as binary input for the build ...\n", clean, commit)
 				if gitErr != nil {
 					return nil, fmt.Errorf("the directory %q is not a valid Git repository: %v", clean, gitErr)
 				}
-				pr, pw := io.Pipe()
+
+				// Create a temp directory to move the repo contents to
+				tempDirectory, err := ioutil.TempDir(os.TempDir(), "oc_cloning_"+options.Commit)
+				if err != nil {
+					return nil, err
+				}
+
 				go func() {
-					if err := repo.Archive(clean, options.Commit, "tar.gz", pw); err != nil {
-						pw.CloseWithError(fmt.Errorf("unable to create Git archive of %q for build: %v", clean, err))
-					} else {
-						pw.CloseWithError(io.EOF)
+					timeout := 300 * time.Second
+					select {
+					case <-transferFinished:
+					case <-time.After(timeout):
+						glog.Warning("Transferring directory %q timed out after %q seconds.", clean, timeout.Seconds())
```

Comment:
is there really a need for this?  the previous iteration of the code didn't worry about the timeout case.  can you just move the removeAll down below where the transfer actually happens, with a flag to indicate if you should removeall (because you cloned to a tempdir) or not removeall (because you just uploaded from a dir the user specified)